File: solutions/24-swap-nodes-in-pairs/swap-nodes-in-pairs.ts
function swapPairs(head: ListNode | null): ListNode | null {
    const dummy = new ListNode(-1);
    dummy.next = head;
    let prevNode = dummy;
    while (head && head.next) {
        const firstNode = head;
        const secondNode = head.next;
        prevNode.next = secondNode;
        firstNode.next = secondNode.next;
        secondNode.next = firstNode;
        prevNode = firstNode;
        head = firstNode.next;
    }
    return dummy.next;
}

File: solutions/24-swap-nodes-in-pairs/swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
    // Dummy node acts as the prevNode for the head node
    // of the list and hence stores pointer to the head node.
    dummy := &ListNode{-1, nil}
    dummy.Next = head
    prevNode := dummy
    for head != nil && head.Next != nil {
        // Nodes to be swapped
        firstNode := head
        secondNode := head.Next
        // Swapping
        prevNode.Next = secondNode
        firstNode.Next = secondNode.Next
        secondNode.Next = firstNode
        // Reinitializing the head and prevNode for next swap
        prevNode = firstNode
        head = firstNode.Next // jump
    }
    // Return the new head node.
    return dummy.Next
}